File: tools/bench/xi3_lost_update.py
"""잃어버린 갱신(lost update) 재현율 측정 — XI-3.

무엇을 재는가
    (1) 전역 정수에 대한 순수 `x += 1` 을 4스레드로 돌렸을 때 유실이 나는가
    (2) 읽기와 쓰기 사이에 함수 호출이 하나 끼었을 때 유실이 나는가
    (3) 스위치 간격(sys.setswitchinterval)에 따라 (2)의 유실률이 어떻게 변하는가
    (4) 락을 걸면 (2)가 어떻게 되는가

    (1)이 멀쩡한 것은 CPython 이 스레드 전환 검사를 특정 바이트코드(루프 뒤로 가는
    점프, 함수 진입)에서만 하기 때문이지 `+=` 가 원자적이어서가 아니다. (2)가 그것을
    보인다.

측정 환경: CLAUDE.md §1-3 (Ubuntu 24.04 / CPython 3.13.12 / 4코어)

사용법:
    python3.13 tools/bench/xi3_lost_update.py
"""

import statistics
import sys
import threading

N_THREADS = 4          # 코어 수와 같게 둔다
N_EACH = 100_000
EXPECT = N_THREADS * N_EACH
ROUNDS = 100

counter = 0
lock = threading.Lock()


def add_one(v):
    return v + 1


def bump_bare():
    global counter
    for _ in range(N_EACH):
        counter += 1                 # LOAD / ADD / STORE — 원자적이 아니다


def bump_call():
    global counter
    for _ in range(N_EACH):
        tmp = counter                # 읽기
        counter = add_one(tmp)       # 함수 호출 뒤 쓰기. 호출 진입에서 GIL 이 풀릴 수 있다


def bump_call_locked():
    global counter
    for _ in range(N_EACH):
        with lock:
            tmp = counter
            counter = add_one(tmp)


def one_run(job):
    global counter
    counter = 0
    ts = [threading.Thread(target=job) for _ in range(N_THREADS)]
    for t in ts:
        t.start()
    for t in ts:
        t.join()
    return counter


def trial(label, job, rounds=ROUNDS):
    vals = [one_run(job) for _ in range(rounds)]
    ratios = [(EXPECT - v) / EXPECT * 100 for v in vals]
    bad = sum(1 for v in vals if v != EXPECT)
    print(f"[{label}] {rounds}회 중 유실 {bad}회")
    print(f"    최종값 중앙값 {int(statistics.median(vals)):,} / 기대 {EXPECT:,} "
          f"(범위 {min(vals):,} ~ {max(vals):,})")
    print(f"    유실률 중앙값 {statistics.median(ratios):.0f}% "
          f"(범위 {min(ratios):.0f} ~ {max(ratios):.0f}%)")
    return bad


def main():
    print(f"python {sys.version.split()[0]} · GIL={sys._is_gil_enabled()} · "
          f"스레드 {N_THREADS} · 코어 4 · 각 {N_EACH:,}회 · {ROUNDS}회 반복")

    for si in (0.005, 0.0001, 0.000001):
        sys.setswitchinterval(si)
        trial(f"순수 += · switchinterval={si}", bump_bare)

    for si in (0.005, 0.0001, 0.000001):
        sys.setswitchinterval(si)
        trial(f"함수 호출 낀 읽기-쓰기 · switchinterval={si}", bump_call)

    sys.setswitchinterval(0.000001)
    trial("함수 호출 + 락 · switchinterval=1e-06", bump_call_locked)


if __name__ == "__main__":
    main()

File: tools/bench/xi3_lost_update.cpp
// 잃어버린 갱신 재현율 측정 (C++) — XI-3.
//
// 같은 카운터를 세 가지로 올린다. 평범한 int(데이터 경쟁 — 표준상 정의되지 않은 동작),
// std::atomic, 뮤텍스. 100회 반복해 최종값의 중앙값·범위와 유실 발생 횟수를 센다.
//
// 측정 환경: CLAUDE.md §1-3 (Ubuntu 24.04 / g++ 13.3.0 -std=c++17 -O2 / 4코어)
// 빌드: g++ -std=c++17 -O2 tools/bench/xi3_lost_update.cpp -o /tmp/xi3 && /tmp/xi3

#include <algorithm>
#include <atomic>
#include <chrono>
#include <cstdio>
#include <mutex>
#include <thread>
#include <vector>
using namespace std;

const int N_THREADS = 4;          // 코어 수와 같게 둔다
const int N_EACH = 1000000;
const long long EXPECT = (long long)N_THREADS * N_EACH;
const int ROUNDS = 100;

volatile long long raw_counter = 0;   // volatile — 컴파일러가 루프를 접지 못하게 한다
atomic<long long> atomic_counter{0};
long long mutex_counter = 0;
mutex mtx;
atomic<bool> go{false};               // 네 스레드를 같은 순간에 출발시킨다

void wait_go() { while (!go.load(memory_order_acquire)) this_thread::yield(); }

void bump_raw() {
    wait_go();
    for (int i = 0; i < N_EACH; i++) raw_counter = raw_counter + 1;   // 데이터 경쟁
}
void bump_atomic() {
    wait_go();
    for (int i = 0; i < N_EACH; i++) atomic_counter.fetch_add(1, memory_order_relaxed);
}
void bump_mutex() {
    wait_go();
    for (int i = 0; i < N_EACH; i++) {
        lock_guard<mutex> g(mtx);
        mutex_counter += 1;
    }
}

long long one_run(void (*job)(), volatile long long& slot) {
    slot = 0;
    go.store(false);
    vector<thread> ts;
    for (int k = 0; k < N_THREADS; k++) ts.emplace_back(job);
    go.store(true, memory_order_release);
    for (auto& t : ts) t.join();
    return slot;
}

long long one_run_mutex() {
    mutex_counter = 0;
    go.store(false);
    vector<thread> ts;
    for (int k = 0; k < N_THREADS; k++) ts.emplace_back(bump_mutex);
    go.store(true, memory_order_release);
    for (auto& t : ts) t.join();
    return mutex_counter;
}

long long one_run_atomic() {
    atomic_counter.store(0);
    go.store(false);
    vector<thread> ts;
    for (int k = 0; k < N_THREADS; k++) ts.emplace_back(bump_atomic);
    go.store(true, memory_order_release);
    for (auto& t : ts) t.join();
    return atomic_counter.load();
}

void report(const char* label, vector<long long> vals) {
    int bad = 0;
    for (long long v : vals) if (v != EXPECT) bad++;
    sort(vals.begin(), vals.end());
    long long med = vals[vals.size() / 2];
    printf("[%s] %d회 중 유실 %d회\n", label, (int)vals.size(), bad);
    printf("    최종값 중앙값 %lld / 기대 %lld (범위 %lld ~ %lld)\n",
           med, EXPECT, vals.front(), vals.back());
    printf("    유실률 중앙값 %.0f%% (범위 %.0f ~ %.0f%%)\n",
           100.0 * (EXPECT - med) / EXPECT,
           100.0 * (EXPECT - vals.back()) / EXPECT,
           100.0 * (EXPECT - vals.front()) / EXPECT);
}

double bench(void (*runner)()) {
    auto t0 = chrono::steady_clock::now();
    runner();
    return chrono::duration<double>(chrono::steady_clock::now() - t0).count();
}

int main() {
    printf("스레드 %d · 코어 4 · 각 %d회 · %d회 반복\n", N_THREADS, N_EACH, ROUNDS);

    vector<long long> a, b, c;
    for (int r = 0; r < ROUNDS; r++) a.push_back(one_run(bump_raw, raw_counter));
    for (int r = 0; r < ROUNDS; r++) b.push_back(one_run_atomic());
    for (int r = 0; r < ROUNDS; r++) c.push_back(one_run_mutex());

    report("평범한 long long", a);
    report("atomic", b);
    report("mutex", c);
    return 0;
}
